File: source/modules/strategy/src/test/sample_data/inputs/combined_request.rb
include_class "org.marketcetera.strategy.ruby.Strategy"
include_class "java.lang.Integer"

class CombinedRequest < Strategy
  def on_start
      @asks = {}
      @bids = {}
      @totalEventCount = 0
  end
  def on_other data
      do_combined_request
  end
  def on_callback data
      requestIDString = get_property "requestID"
      if(get_property("cancelCep") != nil)
          cancel_data_request Integer.parseInt requestIDString
      else
          cancel_data_request Integer.parseInt requestIDString
      end
  end
  def on_ask ask
      record_symbol ask.getSymbol, @asks
      transcribe_collection "ask", @asks
  end
  def on_bid bid
      record_symbol bid.getSymbol, @bids
      transcribe_collection "bid", @bids
  end
  def record_symbol symbol, collection
      if (collection.has_key? symbol)
          collection[symbol] = collection[symbol] + 1
      else
          collection[symbol] = 1
      end
      @totalEventCount += 1
      if(@totalEventCount >= 50)
          set_property "finished", "true"
      end
  end
  def transcribe_collection key, collection
      collection.each {|symbol, value| set_property key + "-" + symbol, (value += 1).to_s }
  end
  def do_combined_request
      symbols = get_property "symbols"
      marketDataSource = get_property "marketDataSource"
      cepSource = get_property "cepSource"
      compressedStatements = get_property "statements"
      if(compressedStatements != nil) 
          statements = compressedStatements.split("#")
          set_property("requestID", Integer.toString(request_processed_market_data(symbols, marketDataSource, statements.to_java(:string), cepSource)))
      else
          set_property("requestID", Integer.toString(request_processed_market_data(symbols, marketDataSource, nil, cepSource)))
      end
  end  
end

File: source/modules/strategy/src/test/sample_data/inputs/CombinedRequest.java
import java.util.HashMap;
import java.util.Map;

import org.marketcetera.event.AskEvent;
import org.marketcetera.event.BidEvent;
import org.marketcetera.strategy.java.Strategy;
import org.marketcetera.marketdata.MarketDataRequest;

/* $License$ */

/**
 * Sample strategy that tests processed market data requests.
 *
 * @author <a href="mailto:colin@marketcetera.com">Colin DuPlantis</a>
 * @version $Id$
 * @since 1.0.0
 */
public class CombinedRequest
        extends Strategy
{
    /**
     * stores bid counts by symbol
     */
    private final Map<String,Integer> bids = new HashMap<String,Integer>();
    /**
     * stores ask counts by symbol
     */
    private final Map<String,Integer> asks = new HashMap<String,Integer>();
    /**
     * counts total events received
     */
    private int totalEventCount = 0;
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onOther(java.lang.Object)
     */
    @Override
    public void onOther(Object inEvent)
    {
        doCombinedRequest();
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onCallback(java.lang.Object)
     */
    @Override
    public void onCallback(Object inData)
    {
        int requestID = Integer.parseInt(getProperty("requestID"));
        if(getProperty("cancelCep") != null) {
            cancelDataRequest(requestID);
        } else {
            cancelDataRequest(requestID);
        }
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onAsk(org.marketcetera.event.AskEvent)
     */
    @Override
    public void onAsk(AskEvent inAsk)
    {
        recordSymbol(inAsk.getSymbolAsString(),
                     asks);
        transcribeCollection("ask",
                             asks);
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onBid(org.marketcetera.event.BidEvent)
     */
    @Override
    public void onBid(BidEvent inBid)
    {
        recordSymbol(inBid.getSymbolAsString(),
                     bids);
        transcribeCollection("bid",
                             bids);
    }
    /**
     * Records the receipt of a symbol.
     *
     * @param inSymbol
     */
    private void recordSymbol(String inSymbol,
                              Map<String,Integer> inCollection)
    {
        Integer count = inCollection.get(inSymbol);
        if(count == null) {
            inCollection.put(inSymbol,
                             1);
        } else {
            inCollection.put(inSymbol,
                             ++count);
        }
        totalEventCount += 1;
        if(totalEventCount >= 50) {
            setProperty("finished",
                        "true");
        }
    }
    /**
     * Writes the given collection into the common storage area.
     *
     * @param inKey a <code>String</code> value to prepend to each stored symbol
     * @param inCollection a <code>Map&lt;String,Integer&gt;</code> value containing the values to be stored
     */
    private void transcribeCollection(String inKey,
                                      Map<String,Integer> inCollection)
    {
        for(String symbol : inCollection.keySet()) {
            setProperty(inKey + "-" + symbol,
                        Integer.toString(inCollection.get(symbol)));
        }
    }
    /**
     * Executes a request for processed market data.
     */
    private void doCombinedRequest()
    {
        String symbols = getProperty("symbols");
        String marketDataSource = getProperty("marketDataSource");
        String compressedStatements = getProperty("statements");
        String[] statements;
        if(compressedStatements != null) {
            statements = compressedStatements.split("#");
        } else {
            statements = null;
        }
        String cepSource = getProperty("cepSource");
        String stringAPI = getProperty("useStringAPI");
        try {
            if(stringAPI != null) {
                setProperty("requestID",
                            Integer.toString(requestProcessedMarketData(MarketDataRequest.newRequest().withSymbols(symbols).fromProvider(marketDataSource).
                                                                          withContent("TOP_OF_BOOK,LATEST_TICK").toString(),
                                                                        statements,
                                                                        cepSource)));
            } else {
                setProperty("requestID",
                            Integer.toString(requestProcessedMarketData(MarketDataRequest.newRequest().withSymbols(symbols).fromProvider(marketDataSource).
                                                                          withContent("TOP_OF_BOOK,LATEST_TICK"),
                                                                        statements,
                                                                        cepSource)));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
